### ISSUE TITLE
feat: Exponential Moving Average (EMA) of model weights

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#211]
## Summary

Model EMA maintains a shadow copy of model weights using exponential moving average. The EMA weights are used for evaluation/inference and consistently produce better accuracy than the raw trained weights.

**Estimated impact**: ~0.5% Top-1 accuracy improvement on ImageNet.

## What EMA Does

After each optimizer step, update shadow weights:
```
ema_param = decay * ema_param + (1 - decay) * model_param
```

Typical `decay = 0.9999` (very slow-moving average).

During evaluation, swap in EMA weights instead of raw weights.

## Proposed API

```java
public final class ModelEMA {
    private final float decay;
    private final Map<Tensor, float[]> shadowParams;  // ema copies
    
    /**
     * @param model  the model to track
     * @param decay  EMA decay rate (default 0.9999)
     */
    public ModelEMA(Sequential model, float decay) {
        this.decay = decay;
        this.shadowParams = new IdentityHashMap<>();
        // Initialize shadow params as copy of current params
        for (var layer : model.parametricLayers()) {
            for (var param : layer.parameters()) {
                shadowParams.put(param, param.data().clone());
            }
        }
    }
    
    /** Update EMA weights. Call after each optimizer.step(). */
    public void update(Sequential model) {
        for (var layer : model.parametricLayers()) {
            for (var param : layer.parameters()) {
                float[] shadow = shadowParams.get(param);
                float[] current = param.data();
                for (int i = 0; i < shadow.length; i++) {
                    shadow[i] = decay * shadow[i] + (1 - decay) * current[i];
                }
            }
        }
    }
    
    /** Swap EMA weights into model for evaluation. */
    public void applyTo(Sequential model) { ... }
    
    /** Restore original weights after evaluation. */
    public void restore(Sequential model) { ... }
}
```

### Integration via Callback

```java
public final class EMACallback implements Callback {
    private final ModelEMA ema;
    
    @Override
    public void onBatchEnd(TrainingContext ctx) {
        ema.update(model);
    }
    
    @Override
    public void onEpochEnd(TrainingContext ctx) {
        // Swap in EMA weights for evaluation
        ema.applyTo(model);
        // ... evaluation happens ...
        ema.restore(model);
    }
}
```

### Usage

```java
var ema = new ModelEMA(model, 0.9999f);

var trainer = new Trainer(model, loss, optimizer)
    .addCallback(new EMACallback(ema));
```

## Acceptance Criteria

- [ ] `ModelEMA` tracks shadow parameters for all parametric layers
- [ ] `update()` applies EMA formula per-parameter
- [ ] `applyTo()` / `restore()` swap weights for evaluation
- [ ] SIMD inner loop for EMA update (vectorizable)
- [ ] EMA state serializable for checkpointing
- [ ] Works with both CPU and GPU tensors
- [ ] Unit test: EMA weights converge toward model weights
- [ ] Unit test: decay=1.0 keeps initial weights, decay=0.0 copies model weights

## References

- [Mean teachers (Tarvainen & Valpola, 2017)](https://arxiv.org/abs/1703.01780)
- [EfficientNet training](https://arxiv.org/abs/1905.11946) — Uses EMA with decay 0.9999
- [PyTorch EMA implementations](https://github.com/rwightman/pytorch-image-models/blob/main/timm/utils/model_ema.py)
